Added mlang tags support to collections and items

Lenetis opened the pull request:
Item and collection names, field titles, field contents, sorting columns, and quite a few other things now support the {mlang} tags.

Example usage:
`Text that will always be displayed{mlang en} text in english{mlang}{mlang pl} tekst po polsku{mlang} always displayed{mlang en} another text in english{mlang}.`

Lenetis committed:
Added mlang tags support to collections and items

Lenetis committed:
Working page title translation!

Also, moved mlang translation function to AppRuntime and AppExtension, removed TranslationExtension and TranslationRuntime files

Lenetis committed:
Using current user locale by default

Lenetis committed:
Working mlangs in collection selection

punfil reviewed (commented):
Looks pretty good, that's a cool feature.

Detected bug:
http://192.168.22.20:81/collections/9155d0a3-1833-47f6-b8f5-7de3a1b33216/items
![obraz](https://github.com/museum-engineering-project/koillectionPG/assets/77413489/e7dbda81-7d56-4452-b564-616af57a40c7)

bug2 (locale is not properly checked for typos/wrong input):
![obraz](https://github.com/museum-engineering-project/koillectionPG/assets/77413489/a941374b-c76d-4a33-afea-6b554925beb8)
![obraz](https://github.com/museum-engineering-project/koillectionPG/assets/77413489/35740f99-3934-47ca-9d6a-96259e2167f8)

bug3:
![obraz](https://github.com/museum-engineering-project/koillectionPG/assets/77413489/480d8799-1c7d-424a-9b9e-855b66fb9308)


Let's suppose I have sth like this (scenario where sth is not translated into all languages) :
{mlang en}ENGTEXT{mlang}{mlang pl}PLTEKST{mlang}
How to display sth in all languages besides en and pl ^?
I can image someone rebasing to new languages coming from upstream, but not adding translation to all items immediately.

Lenetis committed:
Bug fixes

Lenetis commented:
Bug 1: fixed
Bug 2: I don't really see a problem with that. Besides, what could be the solution?
Bug 3: fixed

> Let's suppose I have sth like this (scenario where sth is not translated into all languages) : {mlang en}ENGTEXT{mlang}{mlang pl}PLTEKST{mlang} How to display sth in all languages besides en and pl ^? I can image someone rebasing to new languages coming from upstream, but not adding translation to all items immediately.

That's a good point. I have two ideas:
1. If not a single tag was matched, display english text by default
2. Add {mlang other} tag or something similar, that will be displayed only if no other tag was matched

The second solution would probably be more universal, but OTOH would require a conscious decision from the user. Both should be equally simple to implement. What do you think?

punfil reviewed (commented):
Mostly LGTM, but need to retest once this branch is rebased to dev

punfil commented:
> Bug 1: fixed Bug 2: I don't really see a problem with that. Besides, what could be the solution? Bug 3: fixed
> 
> > Let's suppose I have sth like this (scenario where sth is not translated into all languages) : {mlang en}ENGTEXT{mlang}{mlang pl}PLTEKST{mlang} How to display sth in all languages besides en and pl ^? I can image someone rebasing to new languages coming from upstream, but not adding translation to all items immediately.
> 
> That's a good point. I have two ideas:
> 
>     1. If not a single tag was matched, display english text by default
> 
>     2. Add {mlang other} tag or something similar, that will be displayed only if no other tag was matched
> 
> 
> The second solution would probably be more universal, but OTOH would require a conscious decision from the user. Both should be equally simple to implement. What do you think?



> Bug 1: fixed Bug 2: I don't really see a problem with that. Besides, what could be the solution? Bug 3: fixed
> 
> > Let's suppose I have sth like this (scenario where sth is not translated into all languages) : {mlang en}ENGTEXT{mlang}{mlang pl}PLTEKST{mlang} How to display sth in all languages besides en and pl ^? I can image someone rebasing to new languages coming from upstream, but not adding translation to all items immediately.
> 
> That's a good point. I have two ideas:
> 
>     1. If not a single tag was matched, display english text by default
> 
>     2. Add {mlang other} tag or something similar, that will be displayed only if no other tag was matched
> 
> 
> The second solution would probably be more universal, but OTOH would require a conscious decision from the user. Both should be equally simple to implement. What do you think?

The first one makes more sense to me.
I mean, with that we force user to translate item * to english, BUT probably if we declared mlang other it would contain english anyways.


Second thoughts, what about a setting which tag is the default language for not translated? That might be difficult though.

bartoszLesniewski reviewed (changes requested):
Generally, mlang tags works correctly, but we MUST define some default behaviour when translation for user language doesn't exist. What about using both options that you mentioned?
I mean:
1. User can add `default` parameter in mlang tag, e.g. `{mlang en default}Some text{mlang}`, or even use separate mlang tag (`{mlang default}Some text{mlang}`) if handling two parameters in one tag is too difficult.
2. If default is not specified, english should be used by default

One more suggestion (if it would take a lot of work, you can skip it):
It is possible that user types more spaces between `mlang` and language shortcut or even between `mlang` and `{` by mistake, e.g. `{mlang  pl}Tekst{mlang}` or `{ mlang pl }Tekst{ mlang }`. For now, such cases aren't handled correctly. Is it possible to do it? 

Also, please update other places in source code where mlangs should be used after rebase to dev

Lenetis committed:
Mlang default attribute

Lenetis committed:
Match mlang tags with any number of spaces or tabs

Lenetis committed:
Added mlang support to choice lists and templates

Lenetis commented:
> 1. User can add `default` parameter in mlang tag, e.g. `{mlang en default}Some text{mlang}`, or even use separate mlang tag (`{mlang default}Some text{mlang}`) if handling two parameters in one tag is too difficult.

The default attribute is a good idea. This is what I ended up with:

- `{mlang en}` - will be displayed only if the locale is set to english
- `{mlang default}` - will be displayed only if no other non-default tag was matched
- `{mlang en default}` - will be displayed if the locale is set to english OR if no other non-default tag was matched

If there are multiple `default` tags, all of them will be displayed if not a single non-default tag was matched.

-----

> 2. If default is not specified, english should be used by default

I don't think displaying english is necessary if we have the default attribute. There are valid situations where some text is outside the mlang tags, and they are used e.g. only as a clarification visible in one language

-----

> One more suggestion (if it would take a lot of work, you can skip it): It is possible that user types more spaces between `mlang` and language shortcut or even between `mlang` and `{` by mistake, e.g. `{mlang pl}Tekst{mlang}` or `{ mlang pl }Tekst{ mlang }`. For now, such cases aren't handled correctly. Is it possible to do it?

Done

-----
 
> Also, please update other places in source code where mlangs should be used after rebase to dev

So labels and Dublin Core, or is there something else besides these two?

punfil commented on assets/controllers/select/default_controller.js:
```diff
@@ -1,17 +1,29 @@
 import { Controller } from '@hotwired/stimulus';
 import Translator from "bazinga-translator";
 import { TsSelect2 } from "../../node_modules/ts-select2/dist/core";
-import { htmlStringToDomElement } from "../../js/utils";
```
rebase bug?

punfil commented on assets/controllers/select/default_controller.js:
```diff
@@ -1,17 +1,29 @@
 
     connect() {
-        this.loadSelect()
-    }
```
Why change the scheme? The less we change the less conflicts
Please consider

Lenetis committed:
Fixed rebase bugs

Lenetis committed:
Fixed default tag bug

Lenetis committed:
Added mlang support to DublinCoreXMLGenerator

Lenetis committed:
Added mlang support to labels generator

Lenetis committed:
Added missing translations in labels generation

punfil commented on templates/App/Collection/show.html.twig:
```diff
@@ -105,16 +105,16 @@
                             {% for datum in collection.data %}
-                                <div class="datum-row s12 l5 xl4">
-                                    <span class="label">{{ datum.label }} :</span>
+                                <div class="datum-row col s12 l5 xl4">
```
we need div class `col` here or rebase issue?
Same below

Please consider

punfil reviewed (changes requested):
Found bug:
printable-list
![obraz](https://github.com/museum-engineering-project/koillectionPG/assets/77413489/8671c082-8822-4402-8c29-98f6610af7ae)

and possible bug:
Export as CSV in Tools:
![obraz](https://github.com/museum-engineering-project/koillectionPG/assets/77413489/64d4c189-e9bf-42fc-b18b-262b2694ba02)

Just want to bring attention here, it exports {mlang}... but in other exports like DublinCore we use transMlang filter.

Did you check what happens when scrapers are used?

punfil reviewed (commented):
Bug 3:
![obraz](https://github.com/museum-engineering-project/koillectionPG/assets/77413489/4a42a149-4ff6-4e2d-a3f8-e31a308f4df5)

"Ekwipunek"

punfil reviewed (commented):
Possible bug (Please consider)
Events history:
![obraz](https://github.com/museum-engineering-project/koillectionPG/assets/77413489/03f4cb95-ca49-4eaf-985a-a74bdc735991)

punfil reviewed (commented):
Bug 4:
![obraz](https://github.com/museum-engineering-project/koillectionPG/assets/77413489/15ae63a1-0aa6-4b7e-a981-4460bc5d3ef8)

punfil reviewed (commented):
I don't know if we want to support it here though:
![obraz](https://github.com/museum-engineering-project/koillectionPG/assets/77413489/97b49489-70dc-4d60-b16f-1349544562de)

punfil reviewed (commented):
Bug 6:
![obraz](https://github.com/museum-engineering-project/koillectionPG/assets/77413489/9055db2e-e817-4007-84d6-c0070bd29e2e)

punfil reviewed (commented):
I don't know if we want to support it here though (album):
![obraz](https://github.com/museum-engineering-project/koillectionPG/assets/77413489/1f9eaf13-f88c-4b8e-adfd-10ac843d8244)

punfil reviewed (commented):
I don't know if we want to support it here though:

![obraz](https://github.com/museum-engineering-project/koillectionPG/assets/77413489/d1eb543b-2ed9-4227-9340-3c608e9e6b17)

punfil reviewed (commented):
![obraz](https://github.com/museum-engineering-project/koillectionPG/assets/77413489/16dfb91e-3f1e-436f-b639-080f23db12f5)
Adding related items tip shows:
{mlang en}ENGTEXT{mlang}{mlang pl}PLTEKST{mlang}{mlang default}DEFAULTEKST{mlang}
instead of PLTEKST

Lenetis committed:
Fixed rebase bugs

Lenetis committed:
Added mlang support to History, Loans, Albums and Photos

Lenetis committed:
Added mlang support to printable lists

punfil reviewed (approved):
LGTM, thanks for implementing the feedback!

Lenetis committed:
Merge branch 'dev' into mlang-translation
